content/en/docs/concepts/services-networking/network-policies.md

```diff
@@ -312,6 +323,85 @@ The value of the label is the namespace name.
 While NetworkPolicy cannot target a namespace by its name with some object field, you can use the
 standardized label to target a specific namespace.
 
+## Pod lifecycle
+
+1. Isolation rules are applied before allow rules
+
+   A NetworkPolicy is applied to a selected pod, when isolation rules are enforced
+for that pod. At that point pod can never establish an outbound connection or receive
+an inbound connection that is not allowed by the NetworkPolicy.
+Allow rules will be applied eventually after the isolation rules (that includes applying both at the same time).
+
+   In the worst case, a newly created pod may have no network connectivity at all, if 
+isolation rules were already applied, but no allow rules were applied yet. 
+
+2. NetworkPolicy is not guaranteed to be applied to a pod before it is started
+
+   Network plugins should attempt to ensure that all policy rules are correctly
+configured as soon as possible after pod creation, but how soon that actually
+happens may depend on the network plugin implementation, current cluster load, 
+Pod and NetworkPolicy creation order, etc.
+
+   If the NetworkPolicy is not applied to a pod before it is started,
+no isolation rules may be present for some time, and pod may be able to establish
+a connection that is denied by an existing NetworkPolicy.
```

Comment:
This makes it sound like a network plugin might just _choose_ to not apply NetworkPolicies, which isn't what we want to be saying at all. I think maybe you're trying to use "applied" as a term of art, defined in the previous section, but it doesn't really come across that way.

I've generally tried to express this in terms of the network plugin "being aware" / "becoming aware" of the NetworkPolicy. So I'd say:

1. There's no way to know for sure when the network plugin becomes aware of a newly-created NetworkPolicy, and depending on how the plugin is implemented, it may be aware of the policy on some nodes but not others.
2. Once the network plugin is aware of the policy (on a node), any pod (on that node) selected by that policy is guaranteed to be isolated.
3. Every NetworkPolicy will be fully applied to all selected pods eventually, ...

Reply:
OK, I get not wanting to be too plugin-focused, but I don't think the "applied to" language is very clear. Actually, I think a lot of what makes it unclear is that you're not talking about _who_ is doing the applying, so it's easy to interpret as "when the user applies a NetworkPolicy to a pod" etc.

IMO if we are talking about it from the user perspective then the relevant points are:

- If you create a pod when there are pre-existing NetworkPolicies that select that pod, then it is guaranteed that the pod will be "isolated" when it starts, and the pod will never have more network connectivity than it is supposed to. However, in some cases it may take a moment after the pod starts for all of the relevant "allow" rules to come into effect.
- If you create a pod and a NetworkPolicy that selects that pod at approximately the same time, and there were not any pre-existing NetworkPolicies that selected that pod, then it is not guaranteed that the pod will be isolated when it starts, because (regardless of which order you created the two objects in), it might take Kubernetes longer to process the NetworkPolicy creation than the Pod creation.

From the end user's perspective, that's all you need to know.

Reply:
> If you create a pod when there are pre-existing NetworkPolicies that select that pod, then it is guaranteed that the pod will be "isolated" when it starts, and the pod will never have more network connectivity than it is supposed to. However, in some cases it may take a moment after the pod starts for all of the relevant "allow" rules to come into effect.

This only holds if you have installed a compatible network plugin. Plenty of clusters use a network plugin where that statment doesn't hold. Specifically, these parts would be false:

Reply:
End users _do_ need to know this caveat!

Reply:
Right, that part is important too but I think it probably belongs earlier on in the docs. If you get this far, we're assuming you actually have working NP, I think

Reply:
> If you create a pod when there are pre-existing NetworkPolicies that select that pod

I think we've discussed it before that it is impossible to say which policies are "pre-existing", more specifically it means that the policies are created some time before the pod is created, and then we can't define how much time it should be exactly. While there is no status on network policy from the plugin saying that this policy was handled, we don't know when that happens.

That is generally why this section is so complicated.

Reply:
I restructured this paragraph to focus more on what is guaranteed after the network policy is handled by the network plugin, @danwinship lmk what you think about this version